### PR TITLE
Add an npm package output

### PR DESCRIPTION
**Pull request by alexeagle**
/cc @achew22

### CODE REVIEW

File: BUILD [hunk 1]
```diff
@@ -54,3 +55,48 @@ go_binary(
     embed = [":go_default_library"],
     visibility = ["//visibility:public"],
 )
+
+
+go_binary(
+    name = "bazelisk-darwin",
+    out = "bazelisk-darwin_amd64",
+    embed = [":go_default_library"],
+    goarch = "amd64",
```
Comment (philwo) [inline BUILD]: I didn't know that one could create platform-specific go_binary targets - that's much better than my existing approach in `build.sh`! :) Thanks!

File: bazelisk.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  const extension = {
+    'darwin' : '',
+    'linux' : '',
+    'win32' : '.exe',
+  }[os.platform()];
+
+  if (arch == undefined || platform == undefined) {
+    console.error(`FATAL: Your platform/architecture combination ${
+        os.platform()} - ${os.arch()} is not yet supported.
+    Follow install instructions at https://github.com/bazelbuild/bazel-watcher/blob/master/README.md to compile for your system.`);
```
Comment (philwo) [inline bazelisk.js]: How are the bazel-watcher install instructions related to building Bazel on an unsupported architecture? 🤔

File: bazelisk.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  ps.on('close', e => process.exitCode = e);
+}
+
+if (require.main === module) {
+  main(process.argv.slice(2));
+}
+
+module.exports = {
+  getNativeBinary,
+};
```
Comment (philwo) [inline bazelisk.js]: nit: add trailing newline?

File: build.sh [hunk 4]
```diff
@@ -44,3 +41,10 @@ rm -f bazelisk
 ### Print some information about the generated binaries.
 ls -lh bin/*
 file bin/*
+
+# Non-googlers: you should run this script with NPM_REGISTRY=https://registry.npmjs.org
+readonly REGISTRY=${NPM_REGISTRY:-https://wombat-dressing-room.appspot.com}
+# Googlers: you should npm login using the go/npm-publish service:
+# $ npm login --registry https://wombat-dressing-room.appspot.com
+# TODO: also publish the release to npm.
+# ./bazelisk run --config=release //:npm_package.publish -- --access=public --tag latest --registry $REGISTRY
```
Comment (philwo) [inline build.sh]: Does the "TODO" mean that this is a manual step, or is there still something missing before we can publish to npm?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "name": "@bazel/bazelisk",
+    "version": "0.0.0-PLACEHOLDER",
```
Comment (philwo) [inline package.json]: How is the "version" here replaced with the correct version?

File: rules_nodejs.patch [hunk 6]
```diff
@@ -0,0 +1,13 @@
+diff internal/pkg_npm/packager.js internal/pkg_npm/packager.js
```
Comment (philwo) [inline rules_nodejs.patch]: What does the patch do?

File: BUILD (hunk 1 above)
Comment (alexeagle) [reply BUILD]: Yeah I figured you might want to refactor that script too. If nothing else, this allows you to build all the architectures in parallel, even if you still loop through them to place them in a `bin` directory

File: bazelisk.js (hunk 2 above)
Comment (alexeagle) [reply bazelisk.js]: Oops this is some leftover copy-pasta. Fixed

File: build.sh (hunk 4 above)
Comment (alexeagle) [reply build.sh]: It just needs some manual testing. I could do that myself, but maybe it's better for you or Florian to do the first release to make sure permissions are right.

File: package.json (hunk 5 above)
Comment (alexeagle) [reply package.json]: It's a feature of the pkg_npm rule: https://github.com/bazelbuild/rules_nodejs/blob/master/docs/Built-ins.md#replace_with_version

File: rules_nodejs.patch (hunk 6 above)
Comment (alexeagle) [reply rules_nodejs.patch]: I've never used pkg_npm in a repo that uses `v` prefix on the tags. NPM package versioning has to be purely numeric so it needs to be stripped. I'll make the same change upstream.

File: build.sh (hunk 4 above)
Comment (alexeagle) [reply build.sh]: I removed the comment, but it means after merging this, you'll trip over any problems with npm publishing the next time you run it.

(Also I notice the `build.sh` script doesn't actually do a publish step yet, so maybe this isn't the right place for this command to be added? I didn't see any other release instructions where it belongs instead)

File: stamp.sh [hunk 7]
```diff
@@ -7,3 +7,5 @@ CURRENT_TAG=$(git tag -l --points-at HEAD | head -n1)
 CURRENT_COMMIT=$(git rev-parse HEAD)
 
 echo "STABLE_VERSION ${CURRENT_TAG:-$CURRENT_COMMIT}"
+# rules_nodejs expects to read from volatile-status.txt
+echo "BUILD_SCM_VERSION ${CURRENT_TAG:-$CURRENT_COMMIT}"
```
Comment (achew22) [inline stamp.sh]: I've found it valuable to add a `-dirty` suffix when user changes happen. This saved me trying to debug someone's issue which would have been impossible without it

Comment (alexeagle) [reply stamp.sh]: Agreed, for Angular we use a `.with-local-changes` suffix in a non-pristine clone
https://github.com/angular/angular/blob/master/tools/bazel_stamp_vars.js#L52-L53

But of course that change should be made for all the stamping here, not just npm, so I don't think it should go in this PR. @philwo do you want a separate PR for that?

**Comment by alexeagle** [comment]
Ping! I'd like to build on top of this. If you're okay with it, I could publish the first version of the package from the last release tag.

**Comment by philwo** [comment]
Sorry for the delay, @alexeagle! Merging this now and will try to push a first release.

**Pull request merged**